xen/arch/x86/slaunch.c

```diff
@@ -324,4 +374,52 @@ void tpm_process_drtm_policy(const struct boot_info *bi)
 
         policy_entry[i].flags |= SLR_POLICY_FLAG_MEASURED;
     }
+
+    /*
+     * On x86 EFI platforms Xen reads its command-line options and kernel/initrd
+     * from configuration files (several can be chained). Bootloader can't know
+     * contents of the configuration beforehand without parsing it, so there
+     * will be no corresponding policy entries. Instead, measure command-line
+     * and all modules here.
+     */
+    if ( efi_enabled(EFI_BOOT) )
+    {
+#define LOG_DATA(str) (uint8_t *)(str), (sizeof(str) - 1)
+
+        tpm_hash_extend(DRTM_LOC, DRTM_DATA_PCR,
+                        (const uint8_t *)bi->cmdline, strlen(bi->cmdline),
+                        DLE_EVTYPE_SLAUNCH, LOG_DATA("Xen's command line"));
+
+        for ( i = 0; i < bi->nr_modules; i++ )
+        {
+            const struct boot_module *mod = &bi->mods[i];
+
+            paddr_t string = mod->cmdline_pa;
+            paddr_t start = mod->start;
+            size_t size = mod->size;
+
+            if ( mod->relocated || mod->released )
+            {
+                panic("Multiboot module \"%s\" (at %d) was consumed before measurement\n",
+                      (const char *)__va(string), i);
+            }
+
+            /*
+             * Measuring module's name separately because its command-line
+             * parameters are appended to its name when present.
+             *
+             * 2 MiB is minimally mapped size and it should suffice.
+             */
+            map_l2(string, 2 * 1024 * 1024);
+            tpm_hash_extend(DRTM_LOC, DRTM_CODE_PCR,
+                            __va(string), strlen(__va(string)),
+                            DLE_EVTYPE_SLAUNCH, LOG_DATA("MB module string"));
+
+            map_l2(start, size);
+            tpm_hash_extend(DRTM_LOC, DRTM_DATA_PCR, __va(start), size,
```

Comment:
`DRTM_CODE_PCR` and `DRTM_DATA_PCR` are reversed.

Reply:
https://github.com/TrenchBoot/xen/compare/aa2fcc54193471e68cd2a730d5241e919128686b..f484787ce7d2e2a1c0c4c7929ad9d8eeedcf7db3#diff-1d266d1dff9daf02e019100b9c620a099561c98811a05a0b7581363b3c00ac39R414-R419